### DISCUSSION TITLE
[flask-socketio]Deploying on ubuntu

### DISCUSSION THREAD

**Discussion by JoyX2** [miguelgrinberg/Flask-SocketIO#1816] [Q&A]
Hello Miguel,

I made a simple app with socket-io for my class that works flawlessly on my local computer but as soon as I tried to deploy it using online I have realized that it's a whole different story and it's very hard to get it to work.

So I have decided to start up with a much simpler app because I doubt that my code is the issue rather than the right method of deployment,

Server Python-Flask :

```
import eventlet
import eventlet.wsgi
eventlet.monkey_patch()

from flask import Flask, request
from flask_socketio import SocketIO, emit

app = Flask(__name__)
app.config['SECRET_KEY'] = 'Some_Secret!'
socketio = SocketIO(app, CORS_ALLOW_ALL_ORIGINS=True) 

@app.route("/", methods=['GET', 'POST'])
def index():
    return 'Hello Socket-IO'
################ <SOCKETIO> ######################
@socketio.on('connect')
def connect(namespace="socIO"):
    print(f'Server has connected the Client!')
    emit('Server2Client', "msg from the server ")


@socketio.on('Server2Client')
def S2Cfn(m):
    print(f"Server2Client was called")


@socketio.on('disconnect')
def disconnect():
    print('Client disconnected', request.sid)
################################################
if __name__ == '__main__':
    socketio.run(app, debug=True, port=8000)
```

for the client, the aim is to connect socketIO with a chat application made in unity (C#) with the backend "flask" but since Javascript is safer to test the connection, I made this simple script 

```
<!DOCTYPE HTML>
<html>
<head>
    <title>Flask-SocketIO Test</title>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/3.0.4/socket.io.js" integrity="sha512-aMGMvNYu8Ue4G+fHa359jcPb1u+ytAF+P2SCb+PxrjCdO3n3ZTxJ30zuH39rimUggmTwmh2u7wvQsDTHESnmfQ==" crossorigin="anonymous"></script>
    <script type="text/javascript" charset="utf-8">

        $(document).ready(function() {
            console.log("page ready");
            var socket = io('xxxx.xxxx.xxxx.xxxx:8000', { path: '/socio'});
            socket.on('connect', function() {
                socket.emit('my_event', {data: 'I\'m connected!'});
            });
        });

    </script>
</head>
</html>
```

for the unity wrapper i used this packaged from github.
https://github.com/KyleDulce/Unity-Socketio

and I used the a simple connect script.

```
         KyleDulce.SocketIo.Socket s;
        s = SocketIo.establishSocketConnection("ws://xxxxx.xxxxx.xxxxx.xxxx:8000", 'path=/socio');
        s.connect();
        s.on("S2C", callServerAgain);
      
       public void callServerAgain(string m)
       {
             Debug.Log("send this to server: " + m);
       }
```

#### 

Now lets talk about the system:
System = Ubuntu 20.04
pip :

```
attrs==19.3.0
Automat==0.8.0
bidict==0.22.0
blinker==1.4
certifi==2019.11.28
chardet==3.0.4
click==8.1.2
cloud-init==22.1
colorama==0.4.3
command-not-found==0.3
configobj==5.0.6
constantly==15.1.0
cryptography==2.8
dataclasses==0.6
dbus-python==1.2.16
distro==1.4.0
distro-info===0.23ubuntu1
dnspython==1.16.0
entrypoints==0.3
eventlet==0.30.2
Flask==2.1.1
Flask-SocketIO==5.1.1
Flask-SQLAlchemy==2.5.1
greenlet==1.1.2
gunicorn==20.0.4
httplib2==0.14.0
hyperlink==19.0.0
idna==2.8
importlib-metadata==4.11.3
incremental==16.10.1
iotop==0.6
itsdangerous==2.1.2
Jinja2==3.1.1
jsonpatch==1.22
jsonpointer==2.0
jsonschema==3.2.0
keyring==18.0.1
language-selector==0.1
launchpadlib==1.10.13
lazr.restfulclient==0.14.2
lazr.uri==1.0.3
MarkupSafe==2.1.1
more-itertools==4.2.0
netifaces==0.10.4
oauthlib==3.1.0
passlib==1.7.4
pexpect==4.6.0
psycopg2-binary==2.9.3
pyasn1==0.4.2
pyasn1-modules==0.2.1
PyGObject==3.36.0
PyHamcrest==1.9.0
PyJWT==1.7.1
pymacaroons==0.13.0
PyNaCl==1.3.0
pyOpenSSL==19.0.0
pyrsistent==0.15.5
pyserial==3.4
python-apt==2.0.0+ubuntu0.20.4.7
python-debian===0.1.36ubuntu1
python-engineio==4.3.1
python-socketio==5.5.2
PyYAML==5.3.1
requests==2.22.0
requests-unixsocket==0.2.0
SecretStorage==2.3.1
service-identity==18.1.0
simplejson==3.16.0
six==1.14.0
sos==4.3
SQLAlchemy==1.4.35
ssh-import-id==5.10
systemd-python==234
Twisted==18.9.0
ubuntu-advantage-tools==27.7
ufw==0.36
unattended-upgrades==0.1
urllib3==1.25.8
wadllib==1.3.3
Werkzeug==2.1.1
zipp==1.0.0
zope.interface==4.7.1
```

system has Nginx and gunicorn installed in it.

Nginx:

```
server {
    listen 80;
    server_name xxxx.xxxx.xxxx.xxxx;


    location /static {
        alias /home/root/NTZ1/static;
    }

    location / {
        include /etc/nginx/proxy_params;
        proxy_pass http://127.0.0.1:8000;
        proxy_redirect off;
    }

    location /socio {
        include /etc/nginx/proxy_params;
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_pass http://127.0.0.1:8000/socio;
    }
}

```

I didnt do anything to Gunicorn, if I'm not wrong there was no mention of any gunicorn configuration in the flask-socketio documentation.

Errors that I've received are mostly coming from Nginx itself.

that is how I start the server 
```
 gunicorn --worker-class eventlet  -w 3 socTest:app
```
```
[alert] 49598#49598: 768 worker_connections are not enough
[error] 49598#49598: *99280 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 127.0.0.1, server:xxx.xxx.xxx.xxx, request: "GET /socket.io/?EIO=4&transport=websocket HTTP/1.0", upstream: "http://127.0.0.1:80/socket.io/?EIO=4&transport=websocket", host: "xxx.xxx.xxx.xxxx:80"

 [error] 2113#2113: *2575 connect() failed (111: Connection refused) while connecting to upstream, client: xx.xxx.xxx.xxx, server: xxx.xxxx.xxx.xxx, request: "GET /.git/HEAD HTTP/1.0", upstream: "http://127.0.0.1:8000/.git/HEAD"
````
please let me know if u need anything, I'm almost available all the time.
Thank you

**Comment by miguelgrinberg** [comment]
There is a complete section dedicated to Gunicorn: https://flask-socketio.readthedocs.io/en/latest/deployment.html#gunicorn-web-server

**Comment by JoyX2** [comment]
> There is a complete section dedicated to Gunicorn: https://flask-socketio.readthedocs.io/en/latest/deployment.html#gunicorn-web-server

thank you for replying , Yes I read that section , that is where i got this command from 
` gunicorn --worker-class eventlet  -w 3 socTest:app` and ` gunicorn --worker-class eventlet  -w 1 socTest:app`

i didn't give gevent a try as I have a very simple application, and I don't think eventlet is the issue.

i'm getting ` connect() failed (111: Connection refused) while connecting to upstream,` by the 100s in just a few seconds, what does it means ? why is the connection being refused?

**Comment by miguelgrinberg** [comment]
Sorry but you are going to have to make an effort explaining the issue in more detail.

First you said you did not see any references to gunicorn in the documentation. After I pointed you to it, you claim that you've found two commands there, even though the docs show only one for eventlet. Which of the two are you using? The `-w 1` is the correct one, `-w 3` is not a supported setting.

You also mention you are getting an error. This is too vague. Where is this error? Client? server? Maybe you have a proxy in between? What is the HTTP status code of the error? Are there any logs in the server regarding this error? See the troubleshooting section of the docs to learn how to generate Socket.IO logs.

**Comment by JoyX2** [comment]
*i'm sorry, I'll try to be more detailing, I have decided to make a clean ubuntu 20.04 installation with nothing on it except those i need for the socketio (python3,pip,flask, nginx, eventlet== 0.30.2, Flask_Cors,flask_socketio and gunicorn )

*the data in this comment might be a lot different and that is due to the clean installation*

> First you said you did not see any references to gunicorn in the documentation. After I pointed you to it, you claim that you've found two commands there, even though the docs show only one for eventlet. Which of the two are you using? The `-w 1` is the correct one, `-w 3` is not a supported setting.

*The hosting service suggested that I use `-w 3 ` based on cpu resources available on the hosting server and socketio doc suggested `-w 1` so I gave both a try and both of them have failed resulting the same error I've mentioned , but I'm sticking with 1 worker now.*
```
 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 127.0.0.1, server:xxx.xxx.xxx.xxx, request: "GET /socket.io/?EIO=4&transport=websocket HTTP/1.0", upstream: "http://127.0.0.1:80/socket.io/?EIO=4&transport=websocket", host: "xxx.xxx.xxx.xxxx:80"

```
> You also mention you are getting an error. This is too vague. Where is this error? Client? server?
 
 *the error I've mentioned above is on the Server ( Ngnix error.log )*
 
 *on chrome there is a different error*
 ```
websocket.js:82 WebSocket connection to 'ws://xxx.xxx.xxx.xxx/socket.io/?EIO=4&transport=websocket' failed: 
 ``` 
 >Maybe you have a proxy in between? What is the HTTP status code of the error? Are there any logs in the server regarding this error? See the troubleshooting section of the docs to learn how to generate Socket.IO logs.
 
 *I've modified the Ngnix a little bit, just allowing all origins but that is all I have regarding the server setup*
 ```
server {
	
    add_header Access-Control-Allow-Origin *;
    listen 80;
    server_name xxx.xxx.xxx.xxx;
	
    location / {
        proxy_pass http://127.0.0.1:8000$1$is_args$query_string;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

        proxy_set_header X-Real-IP $remote_addr;

        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_http_version 1.1;

        proxy_set_header Upgrade $http_upgrade;

        proxy_set_header Connection "upgrade";
    }
}
 
 ```
 
> What is the HTTP status code of the error? 

```
![image](https://user-images.githubusercontent.com/101886922/163655040-fd2f85b9-f8e8-4da3-a3ce-3dbf4ce9268e.png)

```
  
  >Are there any logs in the server regarding this error? See the troubleshooting section of the docs to learn how to generate Socket.IO logs
  
  `curl "https://example.com/socket.io/?EIO=4&transport=polling"`
  results:
  
![image](https://user-images.githubusercontent.com/101886922/163655087-365f80f6-58c9-4f43-88bd-623f341b4ecc.png)

when i used `localStorage.debug = '*';` on the browser 

![image](https://user-images.githubusercontent.com/101886922/163655302-809ca14e-e749-45a9-9522-1966b8d818e4.png)

**Comment by JoyX2** [comment]
my current SERVER code:
```
import eventlet
import eventlet.wsgi
eventlet.monkey_patch()

from flask import Flask, request
from flask_socketio import SocketIO, emit
from flask_cors import CORS
import  requests

app = Flask(__name__)
app.config['SECRET_KEY'] = 'Some_Secret!'
CORS(app)
socketio = SocketIO(app, CORS_ALLOW_ALL_ORIGINS=True) 

@app.route("/", methods=['GET', 'POST'])
def index():
    return 'Hello Socket-IO'

@socketio.on('connect')
def connect():
    print(f'Server has connected the Client!')
    emit('Server2Client', "msg from the server ")

@socketio.on('Server2Client')
def HandShake_C2S(m):
    print(f"Server2Client ==> {m}")

@socketio.on('disconnect')
def disconnect():
    print('Client disconnected', request.sid)
    
if __name__ == '__main__':
    socketio.run(app,host='0.0.0.0', debug=True)
```
    
Javascript :
    
```
    <!DOCTYPE HTML>
<html>
<head>
    <title>Flask-SocketIO 2</title>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/3.0.4/socket.io.js" integrity="sha512-aMGMvNYu8Ue4G+fHa359jcPb1u+ytAF+P2SCb+PxrjCdO3n3ZTxJ30zuH39rimUggmTwmh2u7wvQsDTHESnmfQ==" crossorigin="anonymous"></script>
    <script type="text/javascript" charset="utf-8">

        $(document).ready(function() {
            console.log("page ready");
            var socket = io('ws://xxx.xxx.xxx.xxx',{
      		transports: ['websocket', 'polling', 'flashsocket'],
    		});
            socket.on('connect', function() {
            	console.log("Client's connected");
            });
        });

    </script>
</head>
</html>
```

**Comment by miguelgrinberg** [comment]
The issue is with your nginx configuration. I recommend that you start from the nginx configuration example shown in the Socket.IO documentation, and only once you have that working start making changes to adapt it to your needs.

**Comment by JoyX2** [comment]
> The issue is with your nginx configuration. I recommend that you start from the nginx configuration example shown in the Socket.IO documentation, and only once you have that working start making changes to adapt it to your needs.

I've tried both configurations on the socketio website 

socket.io
```
server {
    listen 80;
    server_name xx.xxx.xxx.xxx;

    location / {
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header Host $host;

      proxy_pass http://localhost:8000; #port 3000 causes nginx to shut down for some reason.

      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection "upgrade";
    }
  }

```
![image](https://user-images.githubusercontent.com/101886922/163671133-e1d4e1ee-e193-4b67-95a2-6afb0bf01980.png)


then flask-socketIO.io
```
server {
    listen 80;
    server_name xxx.xxxx.xxxx.xxxx;

    location / {
        include proxy_params;
        proxy_pass http://127.0.0.1:5000;
    }

    location /static {
        alias <path-to-your-application>/static;
        expires 30d;
    }

    location /socket.io {
        include proxy_params;
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_pass http://127.0.0.1:5000/socket.io;
    }
}

```
I got the exact same error 
![image](https://user-images.githubusercontent.com/101886922/163671360-8c50b5ca-beaf-40a9-bb40-f916830be9d4.png)

and I've also found the same Nginx error in the error.log 
```
onnect() failed (111: Connection refused) while connecting to upstream, client: <MYCLIENTIP>, server: <MYSERVERIP>, request: "GET /socket.io/?EIO=4&transport=websocket HTTP/1.1", upstream: "http://127.0.0.1:5000/socket.io/?EIO=4&transport=websocket", host: <MYSERVERIP>
```


the Linux server I'm using has no domain connected to it (only public ip address so no SSL as well ), if that has anything todo with it.

I Think it might has to do with the version of socket IO on the client or on the server or even the version of eventlet
is there is any safe versions i could downgrade/upgrade to just for testing?

**Comment by miguelgrinberg** [comment]
I can't really tell you what's wrong, but the error comes from nginx, and says that it cannot connect to your upstream (the Python server).

I asked you before to enable Socket.IO logs in the server and see if there was something interesting there, but you haven't done that. Not sure if this will help though.

You can also try removing nginx and having your client connect directly to the Gunicorn server to see if that provides any more clues. But I really don't know.

**Comment by JoyX2** [comment]
> I can't really tell you what's wrong, but the error comes from nginx, and says that it cannot connect to your upstream (the Python server).
> 
> I asked you before to enable Socket.IO logs in the server and see if there was something interesting there, but you haven't done that. Not sure if this will help though.
> 
> You can also try removing nginx and having your client connect directly to the Gunicorn server to see if that provides any more clues. But I really don't know.

could you tell me how to enable the debug mode on flask-socketio ? 
I already have debug=True on the run command.

In the mean time I'll remove nginx and see what happens.

**Comment by miguelgrinberg** [comment]
https://flask-socketio.readthedocs.io/en/latest/getting_started.html#debugging-and-troubleshooting

**Comment by JoyX2** [comment]
so I've removed ngnix using those two codes
`sudo apt-get purge nginx nginx-common`
then 
`sudo apt-get autoremove`

and that is the error that appeared on the client side ( i think it's the same )
![image](https://user-images.githubusercontent.com/101886922/163672417-4f806531-1c54-421e-9070-7a2c90946c17.png)

adding the debug mode to server side 

```
import eventlet
import eventlet.wsgi
eventlet.monkey_patch()

from flask import Flask, request
from flask_socketio import SocketIO, emit
from flask_cors import CORS
import  requests

app = Flask(__name__)
app.config['SECRET_KEY'] = 'Some_Secret!'
CORS(app)
socketio = SocketIO(app, CORS_ALLOW_ALL_ORIGINS=True) 

@app.route("/", methods=['GET', 'POST'])
def index():
    return 'Hello Socket-IO'

@socketio.on('connect')
def connect():
    print(f'Server has connected the Client!')
    emit('Server2Client', "msg from the server ")


@socketio.on('Server2Client')
def HandShake_C2S(m):
    print(f"Server2Client ==> {m}")


@socketio.on('disconnect')
def disconnect():
    print('Client disconnected', request.sid)


@socketio.on("my error event")
def on_my_event(data):
    raise RuntimeError()

@socketio.on_error_default
def default_error_handler(e):
    print(request.event["message"]) # "my error event"
    print(request.event["args"])    # (data,)

@socketio.on_error()        # Handles the default namespace
def error_handler(e):
    print(request.event["message"]) # "my error event"
    print(request.event["args"])    # (data,)

if __name__ == '__main__':
    socketio.run(app,host='0.0.0.0', debug=True,logger=True,engineio_logger=True)
```
    
I apologize, i'm not sure where are python logs are stored, this is my first time on ubuntu, but google says all logs are under var/log 
    
![image](https://user-images.githubusercontent.com/101886922/163673273-ad440245-3886-4282-9711-3e807997fdf5.png)

I've opened everyone of them , but I couldn't find anyone python related 
those are the last few lines on the syslog
    
```
Apr 16 11:19:20 localhost systemd[1]: session-130.scope: Succeeded.
Apr 16 11:20:25 localhost systemd[1]: Started Session 132 of user root.
Apr 16 11:21:27 localhost systemd[1]: session-132.scope: Succeeded.
Apr 16 11:23:45 localhost systemd[1]: Started Session 133 of user root.
Apr 16 11:25:01 localhost CRON[62280]: (root) CMD (command -v debian-sa1 > /dev/null && debian-sa1 1 1)
Apr 16 11:25:26 localhost systemd[1]: session-133.scope: Succeeded.
Apr 16 11:29:57 localhost systemd[1]: Started Session 135 of user root.
```

**Comment by miguelgrinberg** [comment]
I'm sorry but I don't think I can't really help you with this. You need someone that can teach you how to do this step by step, really the problem here is your lack of experience. When I asked you to remove nginx, that does not mean uninstalling it, it means that your client needs to connect directly to your Gunicorn server, which is on port 5000.

> I apologize, i'm not sure where are python logs are stored,

The logs from the Python process are where you sent them. I don't know what you are doing there, it is up to you to store your logs.

**Comment by JoyX2** [comment]
> I'm sorry but I don't think I can't really help you with this. You need someone that can teach you how to do this step by step, really the problem here is your lack of experience. When I asked you to remove nginx, that does not mean uninstalling it, it means that your client needs to connect directly to your Gunicorn server, which is on port 5000.

*I tried that, it made no difference* 
```
$(document).ready(function() {
            console.log("page ready");
            var socket = io('ws://xxx.xxx.xxx.xxx:5000',{
      		transports: ['websocket', 'polling', 'flashsocket'],
    		});

            socket.on('connect', function() {
            	console.log("I'M CONNECTED now");
            });
        });
```


> > I apologize, i'm not sure where are python logs are stored,
> 
> The logs from the Python process are where you sent them. I don't know what you are doing there, it is up to you to store your logs.

I'm sorry for inconvenient, Yes I'm new I'm still learning but I'm fast learner, plz don't give up on me :D
I thought ubuntu has a system like cpanel that loggs all the logs to a file (stdrr.log) but I was wrong
so I made my own logging system
I've added this 
```
class LoggerWriter:
    def __init__(self, level):
        self.level = level

    def write(self, message):
        if message != '\n':
            self.level(message)

    def flush(self):
        self.level(sys.stderr)
log = logging.getLogger('foobar')
sys.stdout = LoggerWriter(log.debug)
sys.stderr = LoggerWriter(log.warning)

logging.basicConfig(
        level=logging.DEBUG,
        format='%(asctime)s:%(levelname)s:%(name)s:%(message)s',
        filename='appfolder/pylogs.log',
        filemode='a'
        )
```

and yes this time i caught something interesting.

```
ERROR:engineio.server:null is not an accepted origin. (further occurrences of this error will be logged with level INFO)
```

I have also reinstalled Nginx and used those settings 

```
server {
    listen 80;
    server_name xxx.xxx.xxx.xxx;

    location / {
        include proxy_params;
        proxy_pass http://127.0.0.1:5000;
    }

    location /static {
        alias <path-to-your-application>/static;
        expires 30d;
    }

    location /socket.io {
        include proxy_params;
        proxy_http_version 1.1;
        proxy_buffering off;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_pass http://127.0.0.1:5000/socket.io;
    }
}
```

**Reply by miguelgrinberg** [reply]
A null origin occurs because you are probably running your client from a local file instead of the web server.

Move your client to your web server (preferred option), or else configure CORS in your Socket.IO server appropriately. See https://flask-socketio.readthedocs.io/en/latest/api.html#flask_socketio.SocketIO, in particular the `cors_allowed_origin` option.

**Reply by JoyX2** [reply]
> A null origin occurs because you are probably running your client from a local file instead of the web server.
> 
> Move your client to your web server (preferred option), or else configure CORS in your Socket.IO server appropriately. See https://flask-socketio.readthedocs.io/en/latest/api.html#flask_socketio.SocketIO, in particular the `cors_allowed_origin` option.

Heyyy, I love you , thank you so much !! :D

it actually worked this time !

I have 1 more concern though, is there is anyway to optimize how many requests per minute ? 
i mean the server is supposed to handle 100-5000 connection at a time, and correct me if i'm wrong, the server keeps track of the online users by sending a msg/ping to each client every (X) seconds.

If I could modify the value of X say instead of 1 ( every second ) I can make it ( 10-60 seconds ) the server would be more optimized and therefore more stable during heavy traffics

**Reply by miguelgrinberg** [reply]
> is there is anyway to optimize how many requests per minute ?

How many clients can the server handle now? Also, your code (if it is still what you have above) doesn't do anything. So what do you want to optimize?

**Reply by JoyX2** [reply]
> > is there is anyway to optimize how many requests per minute ?
> 
> How many clients can the server handle now? Also, your code (if it is still what you have above) doesn't do anything. So what do you want to optimize?

yes , in short:

  - allow the members to host their room and keep a link for the room on my database ( as long as the host is online, if he disconnect, the room should go down with him or another should be formed with a different host from the same room )

 -as well as if someone disconnect from the room I should remove him from the chat list 

 - if someone send a chat as well, i would like to propagate it to the room's member only.

 - so you could expect me connecting it to a database with 2-3 tables ( rooms , members in the room ) 
I don't need registration here, i was planning to use the SID as identification

I guess this gives a rough idea on the long polling I'm trying to achieve.

**Reply by miguelgrinberg** [reply]
You haven't answered my question. Optimizing means that you have an application that can do X units of work, while you need it to do Y units of work. So you need to figure out how to go from X to Y. You are asking about optimizing something that doesn't exist yet. Build it, measure it, then you can decide if you need to optimize it for your goals.

**Reply by JoyX2** [reply]
> You haven't answered my question. Optimizing means that you have an application that can do X units of work, while you need it to do Y units of work. So you need to figure out how to go from X to Y. You are asking about optimizing something that doesn't exist yet. Build it, measure it, then you can decide if you need to optimize it for your goals.

sounds Like a plan I'll do that, I just thought instead of bothering you again :D, i could try and miss with variable (X) in my example that I've mentioned earlier to build it form the start as durable and stable as possible 

>i mean the server is supposed to handle 100-5000 connection at a time, and correct me if i'm wrong, the server keeps track of the >online users by sending a msg/ping to each client every (X) seconds.
>
>If I could modify the value of X say instead of 1 ( every second ) I can make it ( 10-60 seconds ) the server would be more optimized >and therefore more stable during heavy traffics
